docs/features/incremental-generators.md

```diff
@@ -42,154 +42,711 @@ public class MyGenerator : IIncrementalGenerator { ... }
 An assembly can contain a mix of diagnostic analyzers, source generators and
 incremental generators.
 
-### Initialization
-
-`IIncrementalGenerator` has an `Initialize` method that is called by the host
-(either the IDE or the command-line compiler) exactly once, regardless of the
-number of further compilations that may occur. `Initialize` passes an instance
-of `IncrementalGeneratorInitializationContext` which can be used by the
-generator to register a set of callbacks that affect how future generation
-passes will occur.
-
-Currently `IncrementalGeneratorInitializationContext` supports two callbacks:
+### Pipeline based execution
 
-- `RegisterForPostInitialization(GeneratorPostInitializationContext)`: the same
-  callback in shape and function as supported by Source Generators today
-- `RegisterExecutionPipeline(IncrementalGeneratorPipelineContext)`: replaces Execute, described below
+`IIncrementalGenerator` has an `Initialize` method that is called by the host[^1]
+exactly once, regardless of the number of further compilations that may occur.
 
-### Pipeline based execution
+[^1]: Such as the IDE or the command-line compiler
 
 Rather than a dedicated `Execute` method, an Incremental Generator instead
-creates an execution 'pipeline' as part of the initialization process via the
-`RegisterExecutionPipeline` method:
+defines an immutable execution pipeline as part of initialization. The
+`Initialize` method receives an instance of
+`IncrementalGeneratorInitializationContext`which is used by the generator to
+define a set of transformations:
 
 ```csharp
 public void Initialize(IncrementalGeneratorInitializationContext initContext)
 {
-    initContext.RegisterExecutionPipeline(context =>
-    {
-        // build the pipeline...
-    });
+    // define the execution pipeline here via transformations of initContext
 }
 ```
 
-This pipeline is not directly executed, but instead consists of a set of steps
-that are executed on demand as the input data to the pipeline changes. Between
-each step the data produced is cached, allowing previously calculated values to
-be reused in later computations when applicable, reducing the overall
-computation required between compilations.
+These transformations are not executed directly at initialization. Instead they
+are used to form a directed graph of actions that can be executed on demand later,
+as the input data changes. Between each transformation, the data
+produced is cached, allowing previously calculated values to be re-used where
+applicable. This caching reduces the computation required for subsequent
+compilations. See [caching](#caching) for more details.
```

Comment:
Think this is too early to bring up caching. I would focus at this point in the doc as more of how developers can use the new API. Basically treat it as a "how to use the new API". I would put details like caching in a later section essentially a "why did we set the API up this way". That can be considered more advanced reading. Not everyone needs to know about caching (possibly)

Reply:
I've kept this in for now, as I think its important early on to try and explain the 'why' of the new API and it mostly just links to the more comprehensive section later on. Not attached to it though.